File: src/net/eduard/api/setup/manager/DamagerManager.java
package net.eduard.api.setup.manager;

import java.util.HashMap;
import java.util.Map;

import org.bukkit.Bukkit;
import org.bukkit.entity.Entity;
import org.bukkit.entity.Projectile;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import org.bukkit.event.entity.EntityDamageByEntityEvent;
import org.bukkit.plugin.java.JavaPlugin;

public class DamagerManager implements Listener {
	public static Entity getLastDamager(Entity entity) {
		Entity damager = lastPvP.get(entity);
		if (damager != null) {
			if (damager instanceof Projectile) {
				Projectile projectile = (Projectile) damager;

				if (projectile.getShooter() != null && projectile instanceof Entity) {
					return (Entity) projectile.getShooter();
				}

			}
		}
		return damager;
	}

	private static Map<Entity, Entity> lastPvP = new HashMap<>();
	private static DamagerManager classe = new DamagerManager();
	static {
		Bukkit.getPluginManager().registerEvents(classe, JavaPlugin.getProvidingPlugin(DamagerManager.class));
	}

	@EventHandler
	public void aoIrPvP(EntityDamageByEntityEvent e) {
		lastPvP.put(e.getEntity(), e.getDamager());
	}

}

File: src/main/java/net/eduard/api/lib/manager/DamagerManager.kt
package net.eduard.api.lib.manager

import net.eduard.api.EduardAPI
import org.bukkit.entity.Entity
import org.bukkit.entity.Projectile
import org.bukkit.event.EventHandler
import org.bukkit.event.EventPriority
import org.bukkit.event.entity.EntityDamageByEntityEvent
import org.bukkit.event.player.PlayerQuitEvent

/**
 * Verificador de quem causou o ultimo Dano no jogador
 * @author Eduard
 */
object DamagerManager : EventsManager() {

    private val lastPvP: MutableMap<Entity, Entity> = mutableMapOf()
    private val lastHitTaken = mutableMapOf<Entity, Long>()

    init{
        register(EduardAPI.instance)
    }
    fun getLastDamageMoment(entity: Entity) : Long{
        return lastHitTaken.getOrElse(entity){0}
    }

    fun getLastDamager(entity: Entity): Entity? {
        val damager = lastPvP[entity]?:return null
        if (damager is Projectile) {
            if (damager.shooter != null) {
                return damager.shooter as Entity
            }
        }
        return damager
    }

    @EventHandler(priority = EventPriority.HIGHEST,ignoreCancelled = true)
    private fun onDamage(e: EntityDamageByEntityEvent) {
        lastPvP[e.entity] = e.damager
        lastHitTaken[e.entity] = System.currentTimeMillis()
    }

    @EventHandler(priority = EventPriority.HIGHEST)
    private fun onQuit(e: PlayerQuitEvent) {
        lastPvP.remove(e.player)
        lastHitTaken.remove(e.player)
    }
}